cylc/flow/scripts/profile.py

```diff
@@ -0,0 +1,177 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""cylc profiler [OPTIONS]
+
+Profiler which periodically polls PBS cgroups to track
+the resource usage of jobs running on the node.
+"""
+
+import os
+import re
+import sys
+import time
+import signal
+import subprocess
+from pathlib import Path
+from dataclasses import dataclass
+from cylc.flow.terminal import cli_function
+from cylc.flow.option_parsers import CylcOptionParser as COP
+
+INTERNAL = True
+PID_REGEX = re.compile(r"([^:]*\d{6,}.*)")
+
+
+def get_option_parser() -> COP:
+    parser = COP(
+        __doc__,
+        argdoc=[
+        ],
+    )
+    parser.add_option(
+        "-i", type=int, help="interval between query cycles in seconds",
+        default=10, dest="delay")
+    parser.add_option(
+        "-m", type=str, help="Location of cgroups directory",
+        default="/sys/fs/cgroup",
+        dest="cgroup_location")
+
+    return parser
+
+
+@cli_function(get_option_parser)
+def main(parser, options):
+    """CLI main."""
+    # Register the stop_profiler function with the signal library
+    signal.signal(signal.SIGINT, stop_profiler)
+
+    profile(options)
+
+
+@dataclass
+class Process:
+    """Class for representing CPU and Memory usage of a process"""
+    cgroup_memory_path: str
+    cgroup_cpu_path: str
+
+
+def stop_profiler(*args):
+    """This function will be executed when the SIGINT signal is sent
+     to this process"""
+    print('profiler exited')
+    sys.exit(0)
+
+
+def parse_memory_file(process):
+    """Open the memory stat file and copy the appropriate data"""
+    memory_stats = {}
+
+    for line in open(process.cgroup_memory_path):
+        return int(line) // 1024
+
+
+def parse_cpu_file(process, cgroup_version):
+    """Open the memory stat file and return the appropriate data"""
+    memory_stats = {}
+
+    if cgroup_version == 1:
+        for line in open(process.cgroup_cpu_path):
+            if "usage_usec" in line:
+                return int(re.findall(r'\d+', line)[0]) / 1000
+    elif cgroup_version == 2:
+        for line in open(process.cgroup_cpu_path):
+            # Cgroups v2 uses nanoseconds
+            return int(line) / 1000000
+    else:
+        raise FileNotFoundError("cpu usage files not found")
+
+
+def write_data(data, filename):
+    try:
+        with open(filename, 'w') as f:
+            f.write(data + "\n")
+    except IOError:
+        raise IOError("Unable to write data to file:" + filename)
+
+
+def get_cgroup_dir():
+    """Get the cgroup directory for the current process"""
+    # Get the PID of the current process
+    pid = os.getpid()
+    # Get the cgroup information for the current process
+    result = subprocess.run(['cat', '/proc/' + str(pid) + '/cgroup'], capture_output=True, text=True)
```

Comment:
Look at using `pathlib.Path` for filesystem path building / manipulation.

I think `capture_output` captures both stdout and stderr, so if there is any error here, it will be swallowed. You might want to only capture stdout (`stdout=subprocess.PIPE`) or report stderr if the process fails (`print(result.stderr, file=sys.stderr)`).

Reply:
Updated and now no longer uses subprocess. It's now pure python in a try catch block